src/rabbit_connection_tracking.erl

```diff
@@ -0,0 +1,340 @@
+%%
+
+-module(rabbit_connection_tracking).
+
+%% Abstracts away how tracked connection records are stored
+%% and queried.
+%%
+%% See also:
+%%
+%%  * rabbit_connection_tracking_handler
+%%  * rabbit_reader
+%%  * rabbit_event
+
+-export([boot/0,
+         ensure_tracked_connections_table_for_node/1,
+         ensure_per_vhost_tracked_connections_table_for_node/1,
+         ensure_tracked_connections_table_for_this_node/0,
+         ensure_per_vhost_tracked_connections_table_for_this_node/0,
+         tracked_connection_table_name_for/1, tracked_connection_per_vhost_table_name_for/1,
+         delete_tracked_connections_table_for_node/1, delete_per_vhost_tracked_connections_table_for_node/1,
+         clear_tracked_connection_tables_for_this_node/0,
+         register_connection/1, unregister_connection/1,
+         list/0, list/1, list_on_node/1,
+         tracked_connection_from_connection_created/1,
+         tracked_connection_from_connection_state/1,
+         is_over_connection_limit/1, count_connections_in/1]).
+
+-include_lib("rabbit.hrl").
+
+-import(rabbit_misc, [pget/2]).
+
+%%
+%% API
+%%
+
+-spec boot() -> ok.
+
+%% Sets up and resets connection tracking tables for this
+%% node.
+boot() ->
+  ensure_tracked_connections_table_for_this_node(),
+  rabbit_log:info("Setting up a table for connection tracking on this node: ~p",
+    [tracked_connection_table_name_for(node())]),
+  ensure_per_vhost_tracked_connections_table_for_this_node(),
+  rabbit_log:info("Setting up a table for per-vhost connection counting on this node: ~p",
+    [tracked_connection_per_vhost_table_name_for(node())]),
+  clear_tracked_connection_tables_for_this_node(),
+  ok.
+
+
+-spec ensure_tracked_connections_table_for_this_node() -> ok.
+
+ensure_tracked_connections_table_for_this_node() ->
+  ensure_tracked_connections_table_for_node(node()).
+
+
+-spec ensure_per_vhost_tracked_connections_table_for_this_node() -> ok.
+
+ensure_per_vhost_tracked_connections_table_for_this_node() ->
+  ensure_per_vhost_tracked_connections_table_for_node(node()).
+
+
+-spec ensure_tracked_connections_table_for_node(node()) -> ok.
+
+ensure_tracked_connections_table_for_node(Node) ->
+    TableName = tracked_connection_table_name_for(Node),
+    case mnesia:create_table(TableName, [{record_name, tracked_connection},
+                                         {attributes, record_info(fields, tracked_connection)}]) of
+      {atomic, ok}     -> ok;
+      {aborted, Error} ->
+        rabbit_log:error("Failed to create a tracked connection table for node ~p: ~p", [Node, Error]),
+        ok
+    end.
+
+
+-spec ensure_per_vhost_tracked_connections_table_for_node(node()) -> ok.
+
+ensure_per_vhost_tracked_connections_table_for_node(Node) ->
+    TableName = tracked_connection_per_vhost_table_name_for(Node),
+    case mnesia:create_table(TableName, [{record_name, tracked_connection_per_vhost},
+                                          {attributes, [vhost, connection_count]}]) of
```

Comment:
please also use `record_info` for specifying `tracked_connection_per_vhost` table attributes. otherwise if ever we decide to add more attributes to the table/record, then we'd have to hard code them here as well.